### ISSUE TITLE
Holocene-D: op-batcher: Add Holocene activation awareness

### ISSUE THREAD

**Issue by sebastianst** [ethereum-optimism/optimism#12122]
When Holocene activates, the derivation pipeline state will be mostly reset, [per spec](https://specs.optimism.io/protocol/holocene/derivation.html#activation). So the batcher needs to be aware of this rule in the derivation pipeline, namely:
If the batcher has only partially submitted a channel pre-Holocene and then the L1 timestamp passes the Holocene activation timestamp, it has to submit the first frame(s) of the channel again, because they would be dropped in the frame queue/channel bank at Holocene activation.

**Comment by geoknee** [comment]
@sebastianst do you think if we implement https://github.com/ethereum-optimism/optimism/issues/12124 it might just handle this case?

**Comment by geoknee** [comment]
Answering my own question -- experimenting with #12430 I realise that that feature will have some inherent delay or lag (at least a few L1 block times). This might be too slow for Holocene activation (we know that situation is coming so we can probably respond to it quicker).

**Comment by geoknee** [comment]
This ticket can be closed because the batcher already checks for expected progress, so would detect a channel which straddles activation and recover from it. We will try to confirm this in a kurtosis test. If we find that this mechanism is not sufficient, we can adopt this #12743

**Comment by vdamle** [comment]
Logs from recreate in Kurtosis attached.

* Holocene activation at `block 50`
* Batcher parameters: `--max-channel-duration=60, --max-l1-tx-size-bytes=10000, --target-num-frames=5, --data-availability-type=calldata, --batch-type=0` (`calldata`)
* Batcher image: `1.10.0-rc3`
* * Kurtosis package network params

```
optimism_package:
  chains:
    - participants:
      - el_type: op-geth
        cl_type: op-node
      - el_type: op-reth
        el_image: "op-reth:v1.1.3"
        cl_type: op-node
      network_params:
        fjord_time_offset: 0
        granite_time_offset: 0
        holocene_time_offset: 100
      batcher_params:
        image: us-docker.pkg.dev/oplabs-tools-artifacts/images/op-batcher:1.10.0
        extra_params: [--max-channel-duration=60, --max-l1-tx-size-bytes=10000, --target-num-frames=5, --data-availability-type=calldata, --batch-type=0]
```

```
[op-batcher-op-kurtosis] t=2024-12-11T15:48:08+0000 lvl=info msg="Handling receipt" id=e092b07d4dfc60b20de2f0242f4db417:2
[op-batcher-op-kurtosis] t=2024-12-11T15:48:08+0000 lvl=info msg="Transaction confirmed" tx_id=e092b07d4dfc60b20de2f0242f4db417:2 tx=0x6ce9e260d89ce1282611903391fedd827be0f0129fc7f044ca6d93fe7392ada6 block=0x79ae6587da1a49c449cabf9b9cca215f34421ce9166e71fb7e4f4e1a7ec9990f:30
[op-batcher-op-kurtosis] t=2024-12-11T15:48:08+0000 lvl=warn msg="Aborting state publishing, max duration exceeded"
[op-batcher-op-kurtosis] t=2024-12-11T15:48:08+0000 lvl=info msg="Publishing transaction" service=batcher tx=0x4e3e4e419cc598cd7041f5eac3e067f9c7de586fd4b062faa59515bca6460322 nonce=8 gasTipCap=5000000000 gasFeeCap=7000000000 gasLimit=180340
[op-batcher-op-kurtosis] t=2024-12-11T15:48:08+0000 lvl=info msg="Transaction successfully published" service=batcher tx=0x4e3e4e419cc598cd7041f5eac3e067f9c7de586fd4b062faa59515bca6460322 nonce=8 gasTipCap=5000000000 gasFeeCap=7000000000 gasLimit=180340 tx=0x4e3e4e419cc598cd7041f5eac3e067f9c7de586fd4b062faa59515bca6460322
[op-batcher-op-kurtosis] t=2024-12-11T15:48:08+0000 lvl=warn msg="sequencer did not make expected progress" syncActions="SyncActions{blocksToPrune: 0, channelsToPrune: 0, clearState: 0x270033a1aa86fe162003e3e6c0b1b1fde7120264b39143ee3c7fd9b2de525747:7, blocksToLoad: &{1 74}}" existingBlock=0x19309a04d5ff7916910a7d076efcb2d5627dfa2a131bdce826a8ef559ced7344:27 safeL2=0x51fe293ae5277fcbd1c051f4f20024d3b5775a88b29ebbbbe98bd46038bd065e:0
[op-batcher-op-kurtosis] t=2024-12-11T15:48:08+0000 lvl=info msg="Loading range of multiple blocks into state" start=1 end=74
[op-batcher-op-kurtosis] t=2024-12-11T15:48:08+0000 lvl=info msg="Added L2 block to local state" block=0x4691ed3762731478aad643e07569bf88aaefec5b8eab0b49ecc9c751b7e021f1:1 tx_count=1 time=1733931942
[op-batcher-op-kurtosis] t=2024-12-11T15:48:08+0000 lvl=info msg="Added L2 block to local state" block=0x74e56b76c8b659560621dc83a9a861b0874fc553fca0bc19dbc27baf9584d790:2 tx_count=1 time=1733931944
[op-batcher-op-kurtosis] t=2024-12-11T15:48:08+0000 lvl=info msg="Added L2 block to local state" block=0x991a85b84507efd845706f7ba80fb4b4eb475de6d89f9c04485efdcc0af8c13e:3 tx_count=1 time=1733931946
[op-batcher-op-kurtosis] t=2024-12-11T15:48:08+0000 lvl=info msg="Added L2 block to local state" block=0xf58b4d098a32ae6ec230c7d8c27b076846e9c54ce4ecfea702f65e8268c1722b:4 tx_count=1 time=1733931948
[op-batcher-op-kurtosis] t=2024-12-11T15:48:08+0000 lvl=info msg="Added L2 block to local state" block=0x39e48b87dcd0b316394c4341ad6f7f7a9a326c9d33835b5f7cbe4bf0cc3b776f:5 tx_count=1 time=1733931950
[op-batcher-op-kurtosis] t=2024-12-11T15:48:08+0000 lvl=info msg="Added L2 block to local state" block=0x6b8718a31a3525c0b5e50f6371d6b53379399af93f792c3caeabd374556525c1:6 tx_count=1 time=1733931952
[op-batcher-op-kurtosis] t=2024-12-11T15:48:08+0000 lvl=info msg="Added L2 block to local state" block=0x2cde4c708e4128ee236ab57a9c2bfadea29ecf739dac480131b40e5054b055a6:7 tx_count=1 time=1733931954
[op-batcher-op-kurtosis] t=2024-12-11T15:48:08+0000 lvl=info msg="Added L2 block to local state" block=0xc60742dccf29c2460155c3fc2bf2d91419fa6e3212ca8c4974fb11bcce2e22a0:8 tx_count=1 time=1733931956
...
[op-batcher-op-kurtosis] t=2024-12-11T15:48:09+0000 lvl=info msg="Added L2 block to local state" block=0x95235be971e6c0cc3e050fc0f7888f55082300f7d7645e4a45f16ed236c4f5e1:72 tx_count=1 time=1733932084
[op-batcher-op-kurtosis] t=2024-12-11T15:48:09+0000 lvl=info msg="Added L2 block to local state" block=0xe0c8786c4f2ac19ad463ca9a4f9614f7f0b03ce7949d42e4ebc7cf2788f10458:73 tx_count=125 time=1733932086
[op-batcher-op-kurtosis] t=2024-12-11T15:48:09+0000 lvl=info msg="Added L2 block to local state" block=0xf48023c2b238769e5bd54734e86e298d983f738abcc38342a7bbfe19d68fe862:74 tx_count=1 time=1733932088
[op-batcher-op-kurtosis] t=2024-12-11T15:48:09+0000 lvl=info msg="Created channel" id=1366952367c88c6be2477af019f2ecca l1Head=0x91b60111fda47a1b1c71f68db67492270d2fea2485c5fda1ba8fe95c2090194b:31 blocks_pending=74 l1OriginLastSubmittedChannel=0x270033a1aa86fe162003e3e6c0b1b1fde7120264b39143ee3c7fd9b2de525747:7 batch_type=0 compression_algo=zlib target_num_frames=5 max_frame_size=9999 use_blobs=false
[op-batcher-op-kurtosis] t=2024-12-11T15:48:09+0000 lvl=info msg="Channel closed" id=1366952367c88c6be2477af019f2ecca blocks_pending=47 num_frames=5 input_bytes=59637 output_bytes=40714 oldest_l1_origin=0x270033a1aa86fe162003e3e6c0b1b1fde7120264b39143ee3c7fd9b2de525747:7 l1_origin=0xd3474989a3d3b6745133a51063067c706dc338ab29468b90806ff53603be8fc2:11 oldest_l2=0x4691ed3762731478aad643e07569bf88aaefec5b8eab0b49ecc9c751b7e021f1:1 latest_l2=0x19309a04d5ff7916910a7d076efcb2d5627dfa2a131bdce826a8ef559ced7344:27 full_reason="channel full: compressor is full" compr_ratio=0.6826969834163354
[op-batcher-op-kurtosis] t=2024-12-11T15:48:09+0000 lvl=info msg="Building Calldata transaction candidate" size=10000
[op-batcher-op-kurtosis] t=2024-12-11T15:48:20+0000 lvl=info msg="Transaction confirmed" service=batcher tx=0x4e3e4e419cc598cd7041f5eac3e067f9c7de586fd4b062faa59515bca6460322 block=0xd295251398270515753cc200c8835a3934cbb327df6a5d05f4c680c64c5d90c1:33 effectiveGasPrice=5020247484
[op-batcher-op-kurtosis] t=2024-12-11T15:48:20+0000 lvl=info msg="Handling receipt" id=e092b07d4dfc60b20de2f0242f4db417:3
[op-batcher-op-kurtosis] t=2024-12-11T15:48:20+0000 lvl=info msg="Transaction confirmed" tx_id=e092b07d4dfc60b20de2f0242f4db417:3 tx=0x4e3e4e419cc598cd7041f5eac3e067f9c7de586fd4b062faa59515bca6460322 block=0xd295251398270515753cc200c8835a3934cbb327df6a5d05f4c680c64c5d90c1:33
[op-batcher-op-kurtosis] t=2024-12-11T15:48:20+0000 lvl=warn msg="transaction from unknown channel marked as confirmed" id=e092b07d4dfc60b20de2f0242f4db417:3
[op-batcher-op-kurtosis] t=2024-12-11T15:48:20+0000 lvl=warn msg="Aborting state publishing, max duration exceeded"
[op-batcher-op-kurtosis] t=2024-12-11T15:48:20+0000 lvl=info msg="Publishing transaction" service=batcher tx=0x205392a32d59a23f8560d473fd3500d3448efe746c7bcea93e80adb7e67df00c nonce=9 gasTipCap=5000000000 gasFeeCap=7000000000 gasLimit=180616
[op-batcher-op-kurtosis] t=2024-12-11T15:48:20+0000 lvl=info msg="Loading range of multiple blocks into state" start=75 end=80
```
[batcher-holocene-activation-awareness.log](https://github.com/user-attachments/files/18099403/batcher-holocene-activation-awareness.log)
[safe-head-progression-holocene-activation-awareness.log](https://github.com/user-attachments/files/18099404/safe-head-progression-holocene-activation-awareness.log)
[sequencer-holocene-activation-awareness.log](https://github.com/user-attachments/files/18099402/sequencer-holocene-activation-awareness.log)
